### DISCUSSION TITLE
MVC에서 view가 model에 의존하는 것은 어떨까요? 장점과 단점은 무엇일까요?

### DISCUSSION THREAD

**Discussion by wonsss** [woowacourse-study/2022-code-review-study#10] [Lotto] [answered]
MVC에서 view가 model에 의존하는 것은 어떨까요? 장점과 단점은 무엇일까요?

**Comment by lokba** [comment, marked as answer]
제 생각을 한번 애기해보겠습니다.

음... 저는 의존은 해야한다고 생각합니다.
MVC패턴에서, View의 역할은 단순히 render하는 역할도 있습니다만, 더 자세하게는 Model의 상태에 따라 변할 수 있는 데이터를 render를 하는 역할도 하고 있다고 생각합니다.


다만, 여기서 문제는 View가 Model에 직접적으로 접근하느냐? 간접적으로 접근하느냐?가 저는 문제라고 생각하는데요.

직접적으로 접근하면, 캡슐화를 위반한다고 생각하기에 올바르지 않다고 생각합니다.
그리하여, 간접적으로 접근하면 좋다고 생각합니다.
예를 간단히 든다면, Controller에게 Model의 데이터를 매개변수로 받으면 저는 무방하다고 생각합니다.


그래서 제 생각을 결론을 애기드리자면,
View가 Model에 의존하는 것은 어찌보면 당연할 수 있는 일이다.

**Reply by wonsss** [reply]
> "Controller에게 Model의 데이터를 매개변수로 받으면 저는 무방하다고 생각합니다."

그러면 View가 Model의 데이터를 Controller로부터 매개변수로 받는 것이 아니라, Model을 호출(의존)하여 직접 데이터를 받는 것은 어떤가요?

**Reply by lokba** [reply]
저는 좋지 않은 방식이라고 생각합니다.

View가 Model에 직접 데이터를 받으면, View의 역할이 1개 더 늘어나게 된 셈입니다.
역할이 늘어났다는 것은 -> (원래는 단순히 매개변수로 받고, render를 하면 되었지만) 고려할 게 더 생겼다. -> 프로덕트가 점점 커지면?

이것을 고려했을때, 저는 View는 단순히 Controller에게 받아서 render하는 역할이 적합하다고 생각합니다. 

마르코는 어떻게 생각하시나요?

**Reply by wonsss** [reply]
저도 록바의 의견에 동의합니다. 
view가 model에 의존하면, model이 변경되거나 문제가 생겼을 때, controller 뿐만 아니라 view에서도 문제를 해결해야 돼서 복잡해져 좋지 않을 것 같네요. 

또한, view가 model로부터 직접 데이터를 받고 싶다면,  'controller'와 'view'는 model에서 단 하나의 인스턴스만을 동일하게 가져와야 하겠네요.  그런데 이러한 구조에서 model의 인스턴스를 추가적으로 새로 생성하고 싶을 때, 의존된 게 많아서 더 복잡하거나 불가능할 수도 있을 것 같아요? 따라서 이처럼 model의 여러 인스턴스를 만들어야 하는 수요가 있는 경우에도 view가 model에 의존하는 방식은 적합하지 않을 것 같네요. 

여러분들은 어떻게 생각하세요?

**Reply by lokba** [reply]
>'controller'와 'view'는 model에서 단 하나의 인스턴스만을 동일하게 가져와야 하겠네요.

이거는 미처 생각하지 못했네요. 좋은 생각 감사합니다 🚘
